File: app/src/main/java/org/michaelbel/moviemade/ui/modules/main/MainActivity.java
package org.michaelbel.moviemade.ui.modules.main;

import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;

import org.michaelbel.moviemade.Moviemade;
import org.michaelbel.moviemade.R;
import org.michaelbel.moviemade.ui.base.BaseActivity;
import org.michaelbel.moviemade.ui.modules.account.AccountFragment;
import org.michaelbel.moviemade.ui.modules.main.fragments.NowPlayingFragment;
import org.michaelbel.moviemade.ui.modules.main.fragments.TopRatedFragment;
import org.michaelbel.moviemade.ui.modules.main.fragments.UpcomingFragment;
import org.michaelbel.moviemade.ui.modules.main.appbar.MainAppBar;
import org.michaelbel.moviemade.ui.modules.search.SearchActivity;
import org.michaelbel.moviemade.ui.modules.settings.SettingsActivity;
import org.michaelbel.moviemade.ui.widgets.bottombar.BottomNavigationBar;
import org.michaelbel.moviemade.ui.widgets.bottombar.BottomNavigationItem;
import org.michaelbel.moviemade.utils.DeviceUtil;
import org.michaelbel.moviemade.utils.SharedPrefsKt;

import javax.inject.Inject;

import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatTextView;
import androidx.appcompat.widget.Toolbar;
import androidx.coordinatorlayout.widget.CoordinatorLayout;
import androidx.core.content.ContextCompat;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.Unbinder;

public class MainActivity extends BaseActivity implements BottomNavigationBar.OnTabSelectedListener {

    public static final int DEFAULT_FRAGMENT = 0;

    private Unbinder unbinder;
    private NowPlayingFragment nowPlayingFragment;
    private TopRatedFragment topRatedFragment;
    private UpcomingFragment upcomingFragment;
    private AccountFragment profileFragment;

    @Inject SharedPreferences sharedPreferences;

    @BindView(R.id.toolbar) Toolbar toolbar;
    @BindView(R.id.app_bar) MainAppBar appBar;
    @BindView(R.id.toolbar_title) AppCompatTextView toolbarTitle;
    @BindView(R.id.bottom_navigation_bar) BottomNavigationBar bottomBar;

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == R.id.item_search) {
            startActivity(new Intent(MainActivity.this, SearchActivity.class));
        } else if (item.getItemId() == R.id.item_settings) {
            startActivity(new Intent(MainActivity.this, SettingsActivity.class));
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        setTheme(R.style.AppThemeTransparentStatusBar);
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        unbinder = ButterKnife.bind(this);
        Moviemade.getComponent().injest(this);

        setSupportActionBar(toolbar);
        toolbarTitle.setText(R.string.app_name);
        appBar.setBackgroundColor(ContextCompat.getColor(this, R.color.transparent20));

        CoordinatorLayout.LayoutParams params = (CoordinatorLayout.LayoutParams) appBar.getLayoutParams();
        params.topMargin = DeviceUtil.INSTANCE.getStatusBarHeight(this);

        nowPlayingFragment = new NowPlayingFragment();
        topRatedFragment = new TopRatedFragment();
        upcomingFragment = new UpcomingFragment();
        profileFragment = new AccountFragment();

        bottomBar.setTabSelectedListener(this);
        bottomBar.setBarBackgroundColor(R.color.primary);
        bottomBar.setActiveColor(R.color.md_white);
        bottomBar.setMode(BottomNavigationBar.MODE_FIXED);
        bottomBar.setBackgroundStyle(BottomNavigationBar.BACKGROUND_STYLE_DEFAULT);
        bottomBar.setFirstSelectedPosition(sharedPreferences.getInt(SharedPrefsKt.KEY_MAIN_FRAGMENT, DEFAULT_FRAGMENT))
            .addItem(new BottomNavigationItem(R.drawable.ic_fire, R.string.now_playing).setActiveColorResource(R.color.accent))
            .addItem(new BottomNavigationItem(R.drawable.ic_star_circle, R.string.top_rated).setActiveColorResource(R.color.accent))
            .addItem(new BottomNavigationItem(R.drawable.ic_movieroll, R.string.upcoming).setActiveColorResource(R.color.accent))
            .addItem(new BottomNavigationItem(R.drawable.ic_account_circle, R.string.account).setActiveColorResource(R.color.accent))
            .initialise();

        if (savedInstanceState == null) {
            startCurrentFragment();
        }
    }

    @Override
    public void onTabSelected(int position) {
        switch (position) {
            case 0:
                startFragment(nowPlayingFragment, R.id.fragment_view);
                break;
            case 1:
                startFragment(topRatedFragment, R.id.fragment_view);
                break;
            case 2:
                startFragment(upcomingFragment, R.id.fragment_view);
                break;
            case 3:
                startFragment(profileFragment, R.id.fragment_view);
                toolbarTitle.setText(null);
                appBar.setBackgroundColor(ContextCompat.getColor(MainActivity.this, R.color.transparent));
                break;
        }

        if (position != 3) {
            sharedPreferences.edit().putInt(SharedPrefsKt.KEY_MAIN_FRAGMENT, position).apply();
            toolbarTitle.setText(R.string.app_name);
            appBar.setBackgroundColor(ContextCompat.getColor(MainActivity.this, R.color.transparent20));
        }
    }

    @Override
    public void onTabReselected(int position) {
        if (position != 3) {
            scrollToTop(position);
        }
    }

    @Override
    public void onTabUnselected(int position) {}

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        String action = intent.getAction();
        String data = intent.getDataString();

        if (Intent.ACTION_VIEW.equals(action) && data != null) {
            profileFragment.presenter.createSessionId(sharedPreferences.getString(SharedPrefsKt.KEY_TOKEN, ""));
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        unbinder.unbind();
    }

    @OnClick(R.id.toolbar)
    void toolbarClick(View v) {
        int position = bottomBar.getCurrentSelectedPosition();
        if (position != 3) {
            scrollToTop(position);
        }
    }

    private void scrollToTop(int position) {
        switch (position) {
            case 0:
                if (nowPlayingFragment.getAdapter().getItemCount() == 0) {
                    nowPlayingFragment.presenter.getNowPlaying();
                } else {
                    nowPlayingFragment.recyclerView.smoothScrollToPosition(0);
                }
                break;
            case 1:
                if (topRatedFragment.getAdapter().getItemCount() == 0) {
                    topRatedFragment.presenter.getTopRated();
                } else {
                    topRatedFragment.recyclerView.smoothScrollToPosition(0);
                }
                break;
            case 2:
                if (upcomingFragment.getAdapter().getItemCount() == 0) {
                    upcomingFragment.presenter.getUpcoming();
                } else {
                    upcomingFragment.recyclerView.smoothScrollToPosition(0);
                }
                break;
        }
    }

    private void startCurrentFragment() {
        int position = sharedPreferences.getInt(SharedPrefsKt.KEY_MAIN_FRAGMENT, DEFAULT_FRAGMENT);

        switch (position) {
            case 0:
                startFragment(nowPlayingFragment, R.id.fragment_view);
                break;
            case 1:
                startFragment(topRatedFragment, R.id.fragment_view);
                break;
            case 2:
                startFragment(upcomingFragment, R.id.fragment_view);
                break;
        }
    }
}

File: app/src/main/java/org/michaelbel/moviemade/ui/modules/main/MainActivity.kt
package org.michaelbel.moviemade.ui.modules.main

import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import androidx.coordinatorlayout.widget.CoordinatorLayout
import androidx.core.content.ContextCompat
import kotlinx.android.synthetic.main.activity_main.*
import org.michaelbel.moviemade.R
import org.michaelbel.moviemade.ui.base.BaseActivity
import org.michaelbel.moviemade.ui.modules.account.AccountFragment
import org.michaelbel.moviemade.ui.modules.main.fragments.NowPlayingFragment
import org.michaelbel.moviemade.ui.modules.main.fragments.TopRatedFragment
import org.michaelbel.moviemade.ui.modules.main.fragments.UpcomingFragment
import org.michaelbel.moviemade.ui.modules.search.SearchActivity
import org.michaelbel.moviemade.ui.modules.settings.SettingsActivity
import org.michaelbel.moviemade.ui.widgets.bottombar.BottomNavigationBar
import org.michaelbel.moviemade.ui.widgets.bottombar.BottomNavigationItem
import org.michaelbel.moviemade.utils.DeviceUtil
import org.michaelbel.moviemade.utils.KEY_MAIN_FRAGMENT
import org.michaelbel.moviemade.utils.KEY_TOKEN
import shortbread.Shortcut

class MainActivity : BaseActivity(), BottomNavigationBar.OnTabSelectedListener {

    companion object {
        private const val DEFAULT_FRAGMENT = 0
    }

    private var nowPlayingFragment: NowPlayingFragment? = null
    private var topRatedFragment: TopRatedFragment? = null
    private var upcomingFragment: UpcomingFragment? = null
    private var profileFragment: AccountFragment? = null

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.item_search) {
            startActivity(Intent(this@MainActivity, SearchActivity::class.java))
        } else if (item.itemId == R.id.item_settings) {
            startActivity(Intent(this@MainActivity, SettingsActivity::class.java))
        }

        return super.onOptionsItemSelected(item)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        setTheme(R.style.AppThemeTransparentStatusBar)
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        setSupportActionBar(toolbar)
        toolbar_title!!.setText(R.string.app_name)
        toolbar!!.setOnClickListener {
            val position = bottom_bar!!.currentSelectedPosition
            if (position != 3) {
                scrollToTop(position)
            }
        }

        app_bar!!.setBackgroundColor(ContextCompat.getColor(this, R.color.transparent20))

        val params = app_bar!!.layoutParams as CoordinatorLayout.LayoutParams
        params.topMargin = DeviceUtil.getStatusBarHeight(this)

        nowPlayingFragment = NowPlayingFragment()
        topRatedFragment = TopRatedFragment()
        upcomingFragment = UpcomingFragment()
        profileFragment = AccountFragment()

        bottom_bar!!.setTabSelectedListener(this)
        bottom_bar!!.setBarBackgroundColor(R.color.primary)
        bottom_bar!!.activeColor = R.color.md_white
        bottom_bar!!.setMode(BottomNavigationBar.MODE_FIXED)
        bottom_bar!!.setBackgroundStyle(BottomNavigationBar.BACKGROUND_STYLE_DEFAULT)
        bottom_bar!!.setFirstSelectedPosition(getSharedPreferences().getInt(KEY_MAIN_FRAGMENT, DEFAULT_FRAGMENT))
            .addItem(BottomNavigationItem(R.drawable.ic_fire, R.string.now_playing).setActiveColorResource(R.color.accent))
            .addItem(BottomNavigationItem(R.drawable.ic_star_circle, R.string.top_rated).setActiveColorResource(R.color.accent))
            .addItem(BottomNavigationItem(R.drawable.ic_movieroll, R.string.upcoming).setActiveColorResource(R.color.accent))
            .addItem(BottomNavigationItem(R.drawable.ic_account_circle, R.string.account).setActiveColorResource(R.color.accent))
            .initialise()

        if (savedInstanceState == null) {
            startCurrentFragment()
        }
    }

    override fun onTabSelected(position: Int) {
        when (position) {
            0 -> startFragment(nowPlayingFragment!!, R.id.fragment_view)
            1 -> startFragment(topRatedFragment!!, R.id.fragment_view)
            2 -> startFragment(upcomingFragment!!, R.id.fragment_view)
            3 -> {
                startFragment(profileFragment!!, R.id.fragment_view)
                toolbar_title!!.text = null
                app_bar!!.setBackgroundColor(ContextCompat.getColor(this@MainActivity, R.color.transparent))
            }
        }

        if (position != 3) {
            getSharedPreferences().edit().putInt(KEY_MAIN_FRAGMENT, position).apply()
            toolbar_title!!.setText(R.string.app_name)
            app_bar!!.setBackgroundColor(ContextCompat.getColor(this@MainActivity, R.color.transparent20))
        }
    }

    override fun onTabReselected(position: Int) {
        if (position != 3) {
            scrollToTop(position)
        }
    }

    override fun onTabUnselected(position: Int) {}

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        val action = intent.action
        val data = intent.dataString

        if (Intent.ACTION_VIEW == action && data != null) {
            profileFragment!!.presenter.createSessionId(getSharedPreferences().getString(KEY_TOKEN, "")!!)
        }
    }

    private fun scrollToTop(position: Int) {
        when (position) {
            0 -> if (nowPlayingFragment!!.adapter.itemCount == 0) {
                nowPlayingFragment!!.presenter.getNowPlaying()
            } else {
                nowPlayingFragment!!.recyclerView.smoothScrollToPosition(0)
            }

            1 -> if (topRatedFragment!!.adapter.itemCount == 0) {
                topRatedFragment!!.presenter.getTopRated()
            } else {
                topRatedFragment!!.recyclerView.smoothScrollToPosition(0)
            }

            2 -> if (upcomingFragment!!.adapter.itemCount == 0) {
                upcomingFragment!!.presenter.getUpcoming()
            } else {
                upcomingFragment!!.recyclerView.smoothScrollToPosition(0)
            }
        }
    }

    private fun startCurrentFragment() {
        val position = getSharedPreferences().getInt(KEY_MAIN_FRAGMENT, DEFAULT_FRAGMENT)

        when (position) {
            0 -> startFragment(nowPlayingFragment!!, R.id.fragment_view)
            1 -> startFragment(topRatedFragment!!, R.id.fragment_view)
            2 -> startFragment(upcomingFragment!!, R.id.fragment_view)
        }
    }

    @Shortcut(id = "favorites", rank = 3, icon = R.drawable.ic_shortcut_favorite, shortLabelRes = R.string.favorites)
    fun showFavorites() {
        startFragment(profileFragment!!, R.id.fragment_view)
        toolbar_title!!.text = null
        app_bar!!.setBackgroundColor(ContextCompat.getColor(this@MainActivity, R.color.transparent))
        bottom_bar!!.selectTab(3)
        startFavorites(profileFragment!!.accountId)
    }

    @Shortcut(id = "watchlist", rank = 2, icon = R.drawable.ic_shortcut_bookmark, shortLabelRes = R.string.watchlist)
    fun showUpcomingMovies() {
        startFragment(profileFragment!!, R.id.fragment_view)
        toolbar_title!!.text = null
        app_bar!!.setBackgroundColor(ContextCompat.getColor(this@MainActivity, R.color.transparent))
        bottom_bar!!.selectTab(3)
        startWatchlist(profileFragment!!.accountId)
    }
}